### DISCUSSION TITLE
Controlling movie quality parameters

### DISCUSSION THREAD

**Discussion by VictorSeven** [zwicker-group/py-pde#202] [Q&A] [answered]
First of all, thanks for sharing this package, it's awesome!

I am trying to produce movies that include stochastic noise. Fluctuations in the field do produce a lot of changes that due to video compression result in very poor quality videos. 

Therefore I would like to increase the dpi and bitrate of the video in order to make the images more clear. [In the docs](https://py-pde.readthedocs.io/en/latest/packages/pde.visualization.movies.html#pde.visualization.movies.movie_multiple) I can see that the base `Movie` class do have options to control these, but apparently, I cannot feed `bitrate` as a parameter to other functions, such as `movie`.  Maybe this can be accessed from `plot_args` but I don't know how to do it.

Documentation is not very clear about how to do this. Some things I have tried:

1. Using a `MemoryStorage` as written in docs, section 3. Result: produces an mp4 file and plays, but I cannot adjust the bitrate.

```
    storage = MemoryStorage()
    tracker = PlotTracker(interval=1.0, movie=mymovie, plot_args={"vmin":0, "vmax": 0.5, "figsize":(10,5)})
    sol = eq.solve(state, t_range=10.0, dt=1e-2, tracker=["progress", storage.tracker(1)])
    mymovie = movies.movie(storage, "pelicula.mp4", dpi=300, plot_args={"vmin":0, "vmax": 0.5})
```

 2. Try to feed a `Movie` to a `PlotTracker`. Result: writes a file, but it cannot be read (I guess because of lack of `MemoryStorage`)
 
```
    mymovie = movies.Movie("pelicula.mp4", dpi=300, bitrate=100)
    tracker = PlotTracker(interval=1.0, movie=mymovie, plot_args={"vmin":0, "vmax": 0.5, "figsize":(10,5)})
    sol = eq.solve(state, t_range=10.0, dt=1e-2, tracker=tracker)
```

I guess there is an easy way to control movie parameters, but at least for me, it is not clear how to do this.  I would appreciate some extra indications on how to increase bitrate and framerate. Thank you very much in advance!

**Comment by david-zwicker** [comment, marked as answer]
I agree that setting the details is a bit more complicated and might not be very well exposed in the current API. Your two approaches are very sensible and are close to working, as I will discuss in detail below.

1) You're right that the simple call to `movies.movie` does not allow you to set the bitrate (and `plot_args`). PR #203 fixes this, so in the future you should be able to use
```python
storage = MemoryStorage()
sol = eq.solve(state, t_range=10.0, dt=1e-2, tracker=["progress", storage.tracker(1)])
mymovie = movies.movie(storage, "pelicula.mp4", plot_args={"vmin": 0, "vmax": 0.5},
                       movie_args={"dpi": 300, "bitrate": 2000})
```

2) The second approach is also feasible, but has one pitfall where the object `mymovie` needs to be closed after writing to it using the `PlotTracker`. You could do this by calling `mymovie.save()` after the last line. However, a better approach might be to use the following construct
```python
with movies.Movie("pelicula.mp4", dpi=300, bitrate=100) as mymovie:
    tracker = PlotTracker(interval=1.0, movie=mymovie, plot_args={"vmin": 0, "vmax": 0.5, "figsize": (10, 5)})
    sol = eq.solve(state, t_range=10.0, dt=1e-2, tracker=tracker)
```

P.S: I converted the issue to a discussion, so it is better accessible in the future

**Reply by VictorSeven** [reply]
Thank you very much for your answer and the super-fast PR! I think the new arguments will be very useful. For now, however, I am using the second solution -which works perfectly too.
